recipes/armadillo/all/conanfile.py

```diff
@@ -0,0 +1,489 @@
+            if platform == "all" or platform == self.settings.os:
+                for option, dependencies in opts.items():
+                    dependencies_in_error = [
+                        dependency
+                        for dependency in dependencies
+                        if not getattr(self.options, dependency)
+                    ]
+                    if getattr(self.options, option) and dependencies_in_error:
+                        raise ConanInvalidConfiguration(
+                            "Option {option} cannot be enabled without enabling {missing}".format(
+                                option=option, missing=", ".join(dependencies_in_error)
+                            )
+                        )
+        # Mutually exclusive options
+        for category, incompatible_opts in self.exclusions.items():
+            conflicts = [x for x in incompatible_opts if getattr(self.options, x)]
+            if len(conflicts) > 1:
+                raise ConanInvalidConfiguration(
+                    "The following options are in conflict: {conflict}. Resolve this conflict by ensuring only one of these options is enabled and try again.".format(
+                        conflict=", ".join(conflicts)
+                    )
+                )
+        if self.settings.compiler == "Visual Studio" and self.options.shared:
+            self.output.warn(
+                "Building a shared library with MSVC is not supported. This may not be successful."
+            )
+
+    def requirements(self):
+        # Optional requirements
+        openblas = "openblas/[>=0.3.10]"
+        hdf5 = "hdf5/[>=1.12.0]"
+        # atlas = "atlas/3.10.3" # Pending https://github.com/conan-io/conan-center-index/issues/6757
+        # superlu = "superlu/5.2.2" # Pending https://github.com/conan-io/conan-center-index/issues/6756
+        # arpack = "arpack/1.0" # Pending https://github.com/conan-io/conan-center-index/issues/6755
+        # flexiblas = "flexiblas/3.0.4" # Pending https://github.com/conan-io/conan-center-index/issues/6827
+
+        if self.options.ARMA_USE_HDF5 and not self.options.USE_SYSTEM_HDF5:
+            # Use the conan dependency if the system lib isn't being used
+            self.requires(hdf5)
+            self.options["hdf5"].shared = self.options.shared
+
+        if self.options.USE_OPENBLAS:
+            self.requires(openblas)
+            # Note that if you're relying on this to build LAPACK, you _must_ have
+            # a fortran compiler installed. If you don't, OpenBLAS will build successfully but
+            # without LAPACK support, which isn't obvious.
+            # This can be achieved by setting the FC environment variable in your conan profile
+            self.options["openblas"].build_lapack = (
+                self.options.ARMA_USE_LAPACK and not self.options.USE_SYSTEM_LAPACK
+            )
+            self.options["openblas"].shared = self.options.shared
```

Comment:
you cannot modify requirements options, because it is not reliable if several packages in the graph try to modify the same package's options (the final values depend on the graph's evaluation order). What you should do instead is raise a ConanaInvalidConfiguration in validate(self) if the options of requirements are not compatible with the options of armadillo. This should solve the current CI error

Reply:
A few questions:

1. Does this apply to _all_ options, or just `shared`? I guess I'm not clear if the `shared` option would be propagated up/down the dependency tree.
2. Does this mean that in order to build the openblas with lapack, the user will have to append `-o openblas:build_lapack=True` to the required build statement? If so, I fear that the CI will fail anyway, because this option is `False` by default and I wouldn't imagine the CI will be testing that option.
3. A quick search of this repo identified a number of examples of requirement options being explicitly set within the recipe as well:
* [msgpack](https://github.com/conan-io/conan-center-index/blob/4417d304f54017b451a1453585046864de2c00ad/recipes/msgpack/all/conanfile.py#L68)
* [paho-mqtt-cpp](https://github.com/conan-io/conan-center-index/blob/4417d304f54017b451a1453585046864de2c00ad/recipes/paho-mqtt-cpp/all/conanfile.py#L43)
* [kealib](https://github.com/conan-io/conan-center-index/blob/4417d304f54017b451a1453585046864de2c00ad/recipes/kealib/all/conanfile.py#L38)
Are you sure that this is what it's complaining about? My read was that perhaps openblas hadn't been built with the `build_lapack=True` option and so there wasn't an existing item in cache for it. Not having this will break basic functionality for armadillo, though, as it has a reasonably strong dependency on lapack (though it will build and operate in a _very_ limited capacity without it)

Reply:
> 1. Does this apply to _all_ options, or just `shared`? I guess I'm not clear if the `shared` option would be propagated up/down the dependency tree.
it applies to all options, and no options (including `shared` is propagated up or down the dependency tree


> 3\. A quick search of this repo identified a number of examples of requirement options being explicitly set within the recipe as well:
indeed, these recipe modify dependencies options (which I did not know/forgot it was done like this on CCI). On the other hand, they also check in validate(self) that the option is correct (except msgpack but it's probably just because this recipe has not been maintained in the last 12 months, and should be corrected).

Conan-center will only build packages with default options. `shared` is an exception as documented in https://github.com/conan-io/conan-center-index/blob/master/docs/supported_platforms_and_configurations.md. If armadillo cannot work with openblas's default option, then you could try to make a PR to openblas's recipe to change it's default option, but it may not be accepted

Reply:
yes, that would be the only option for now to change openblas's default option, if armadillo requires that.
in future, CI may support to build other non-default options besides shared, but not today.
building all the possible options is non-starter, some recipes may have thousands or even literally infinite amount of the valid option combinations (some allow arbitrary strings or number as an options, e.g. custom namespace names).

Reply:
As per a previous comment, I've seen [reason to suspect](https://github.com/conan-io/conan-center-index/issues/4696#issuecomment-897002205) that gfortran is available on the CI machines to build lapack.. I'll raise an issue and PR for openblas to change the default option as I think this is the most desirable course of action.

In the event that this isn't accepted/successful for whatever reason, the next best course of action I think is to set the default build of armadillo to build without lapack support. I'll add guards around the appropriate logic in the example and everything should build. This won't be a very useful build, but it will at least make armadillo available and give users the option to build it in a more useful state if they have a fortran compiler or any of the optional system libraries installed.

Reply:
yes, having armadillo available without some features is still much better than having no armadillo at all.
let's try to change the defaults in lapack, and if it's impossible for whatever reason, let's disable lapack support in armadillo by default

Reply:
As per https://github.com/conan-io/conan-center-index/issues/4696#issuecomment-923973246, it looks like gfortran isn't actually available on the CI runners, and the openblas PR #7362 can't be merged until it is. In the meantime, I'll configure this to build without lapack support.